Add a zoekt-dynamic-indexserver cmd with HTTP interface for indexing

DylanGriffith opened the pull request:
I'm trying to build an integration in GitLab for indexing code in Zoekt and querying it as a backend for our code search engine https://gitlab.com/gitlab-org/gitlab/-/merge_requests/105049 .

## Objectives

In my attempts to implement this I ran into a few features I would have liked to have:

1. I want the project name in Zoekt to be the `id` in the GitLab database so that I can efficiently filter to hundreds or thousands of project ids. Using the full project URL like `r:gitlab.com/gitlab-org/gitlab` results in very long regex strings to be parsed and even constructing a reliable regex to match this was kinda a mess. I'd actually prefer to update the webserver to take an array of ids as an argument so I don't need regexes at all but for now I've just implemented a way to pass the repository name to `zoekt-git-clone`
2. I want the client (GitLab) to control the indexing via push for efficiency (looping through and polling millions of repos seems too slow and wasteful) so I've added an API to the indexserver to allow me to tell it to index a repo
3. I want to control the rollout of which repos are indexed from the GitLab side so I want to pass in the repo URL to the indexserver so that it indexes this repo and this same endpoint can be used for updates
4. I want this endpoint to support local `.git` bare repos as well as remote `.git` URLs so that I can use this locally in tests to trigger indexing without having to run a webserver that can accept a `git clone` request. Luckily the `zoekt-git-clone` already does this so it worked without much effort but also this is part of the reason this needs to be push not pull because I don't want to run the git web server during tests

## Questions

This is a scrappy draft implementation and I'm trying to get early feedback on the direction here. A few questions I have:

1. Would you rather be discussing this in the context of an issue or is this PR an OK place for us to evaluate these objectives and iterate towards a solution?
1. Does a dynamic indexing API like this violate some other design goals of Zoekt? It moves from pull to push but it offers flexibility for deployment and testing environments as the client has full control over the index
2. Does this belong in indexserver, webserver or should it be extracted to another binary that wraps the executables it is calling?
3. I notice that there is a proxy from webserver -> sourcegraph-indexserver already implemented. Would it make sense instead to have indexserver listen on a socket and be proxied from webserver in the same way? It could use the same socket as sourcegraph-indexserver so it will just make that implementation more generic
4. What kind of tests are appropriate here? Lots of the `cmd` don't have tests yet. I could add to `web/e2e_test.go` or build something similar which is end-to-end of both indexing and web

I also recognise my Go code here is not great as I'm not really experienced but I'll probably get some more feedback from other Go experts before polishing this up but I wanted to see if the direction made sense first.

## Example usage:

```
$ zoekt-indexserver -data_dir zoekt/data -index_dir zoekt/index -dynamic_server
```

```
$ curl -XPOST -d '{"RepositoryUrl":"https://gitlab.com/gitlab-org/gitlab-elasticsearch-indexer.git","RepositoryId":"2953390"}' -H 'Content-Type: application/json' http://127.0.0.1:6060/index
```

DylanGriffith committed:
Implement /index endpoint in zoekt-indexserver

DylanGriffith commented on cmd/zoekt-indexserver/main.go:
```diff
@@ -164,6 +166,61 @@ func indexPendingRepos(indexDir, repoDir string, opts *Options, repos <-chan str
 }
 
+type indexRequest struct {
+	RepositoryUrl  string // TODO: Decide if tokens can be in the URL or if we should pass separately
```
I may initially start with public project support only but we do need some way to index private repos. I noticed the current GitLab indexer doesn't pass the token through so I had a separate change for that https://github.com/sourcegraph/zoekt/commit/18b83db289931039cb85503290c46e19816fe4a5 but this doesn't use the gitlab indexer. I also had concerns about adding the token to the URL and how that might end up being logged all over the place. I didn't want to tackle this problem at the same time as the indexing API but the design will need to account for it.

DylanGriffith commented on cmd/zoekt-indexserver/main.go:
```diff
@@ -164,6 +166,61 @@ func indexPendingRepos(indexDir, repoDir string, opts *Options, repos <-chan str
+			http.Error(w, "JSON parser error", http.StatusBadRequest)
+		}
+
+		ctx, cancel := context.WithTimeout(context.Background(), indexTimeout)
```
I think in practice we probably don't want HTTP requests blocking for 1 hour (especially from the GitLab side) so might consider making this async but it might be nice to get some acknowledgement of well-formed request first.

DylanGriffith commented on cmd/zoekt-indexserver/main.go:
```diff
@@ -290,15 +348,19 @@ func main() {
-	if err != nil {
-		log.Fatalf("readConfigURL(%s): %v", opts.mirrorConfigFile, err)
-	}
+	if *dynamicIndexServer {
```
This was easier to implement as "either dynamic webserver" OR "config file with polling" since the 2 seemed like completely different approaches. I think this partly makes me inclined towards a separate binary but then we've gotta come up with a new name 🤷

keegancsmith replied:
Yeah I think if they are that different a new binary is a legitimate approach (it is what we did with zoekt-sourcegraph-indexserver). What you are implementing sounds more like a `zoekt-indexserver` than what we had before. Maybe if you wanna split it up (happy with starting off with same binary), the name for the old functionality is `zoekt-indexd` to take some sort of unix inspiration?

DylanGriffith replied:
This was moved to a separate binary now. I was a little concerned about renaming the old library without causing confusion so we went with `zoekt-dynamic-indexserver`. I agree `zoekt-indexd` is a clearer name for the previous functionality. I'm happy to take your lead on what you think is a reasonable risk in terms of breaking changes but renaming binaries will at least cause some headache for people with automated building and deployment of these binaries so it could just be easier to avoid it.

keegancsmith commented:
Hi @DylanGriffith thanks for the contribution. I'll take a look at the PR on Monday, but for now I'll respond to your questions.

> 1. I want the project name in Zoekt to be the `id` in the GitLab database so that I can efficiently filter to hundreds or thousands of project ids. Using the full project URL like `r:gitlab.com/gitlab-org/gitlab` results in very long regex strings to be parsed and even constructing a reliable regex to match this was kinda a mess. I'd actually prefer to update the webserver to take an array of ids as an argument so I don't need regexes at all but for now I've just implemented a way to pass the repository name to `zoekt-git-clone`

I assume you won't be using the HTML interface if you expect users to be passing in repo ids, instead you will be using an RPC interface? If so for Sourcegraph we have added RPC interfaces, extended the index to take a integer ID as well as extending the query package to efficiently query for sets of IDs. We do a lot of the glue here in zoekt-sourcegraph-indexserver for further pointers. Hopefully that is enough to unblock ya.

> 2. I want the client (GitLab) to control the indexing via push for efficiency (looping through and polling millions of repos seems too slow and wasteful) so I've added an API to the indexserver to allow me to tell it to index a repo

That sounds good. The API will be some sort of webhook, or something you are adding?

> 3. I want to control the rollout of which repos are indexed from the GitLab side so I want to pass in the repo URL to the indexserver so that it indexes this repo and this same endpoint can be used for updates

We have the same thing in Sourcegraph, except we still rely on pull instead of push. We have a bespoke endpoint which returns the set of IDs to index, then another endpoint to get actual config per repo. Also have some stuff so that we only get the set of IDs that have changed/etc.

> 1. Would you rather be discussing this in the context of an issue or is this PR an OK place for us to evaluate these objectives and iterate towards a solution?

PR is perfect.

> 2. Does a dynamic indexing API like this violate some other design goals of Zoekt? It moves from pull to push but it offers flexibility for deployment and testing environments as the client has full control over the index

Nope, we can add commands which go about things in different ways, as long as all the pieces fit together it sounds good to me.

> 3. Does this belong in indexserver, webserver or should it be extracted to another binary that wraps the executables it is calling?

Maybe another binary, kind of like what we did for zoekt-sourcegraph-indexserver. But to be honest, I haven't looked at the code yet so maybe indexserver is indeed the best place.

> 4. I notice that there is a proxy from webserver -> sourcegraph-indexserver already implemented. Would it make sense instead to have indexserver listen on a socket and be proxied from webserver in the same way? It could use the same socket as sourcegraph-indexserver so it will just make that implementation more generic

This proxy is sourcegraph specific and I don't think is needed in your case. We do it due to our unique requirements around service discovery. If this would also suit you, then we can adjust the design a bit, but I imagine you can use the same unix socket.

> 5. What kind of tests are appropriate here? Lots of the `cmd` don't have tests yet. I could add to `web/e2e_test.go` or build something similar which is end-to-end of both indexing and web

Most of the cmds are light weight wrappers around a pkg which is better tested. But that is the theory, in practice some of the commands have grown quite large with minimal test coverage. I don't think we need that many e2e tests around indexing and web tbh, we have quite detailed testing around indexing. I suspect just ensuring that a push updates and index is the level of testing needed from what I have read in the PR desc.

DylanGriffith commented:
> That sounds good. The API will be some sort of webhook, or something you are adding?

We will probably integrate GitLab directly to Zoekt so we won't use our generic webhook implementation but I plan to call Zoekt directly but similar to a webhook. I might consider using RPC eventually but getting started with HTTP seemed easier to me and from what I can tell having an RPC capability added after HTTP endpoint seemed fine.

DylanGriffith commented:
@keegancsmith thanks for the fast response and answers to my questions! Glad to hear that this PR is roughly in the right direction but your response also has helped me realise there might be some more opportunities to do this a little better.

> If so for Sourcegraph we have added RPC interfaces, extended the index to take a integer ID as well as extending the query package to efficiently query for sets of IDs. We do a lot of the glue here in zoekt-sourcegraph-indexserver for further pointers. Hopefully that is enough to unblock ya.

I think I can find some of this code but I think I'm missing the query API for this. If I understand correctly you are [adding `zoekt.repoid` as a git config](https://github.com/sourcegraph/zoekt/blob/930cd1c28917e64c87f0ce354a0fd040877cbba1/cmd/zoekt-sourcegraph-indexserver/index.go#LL102) to the repos. Do all these `zoekt.` git config metadata fields end up in the index somehow? I haven't looked into the indexing/search code much yet but from the web UI I'm not seeing any way to query based on that metadata. If there is a way to query on that metadata then I think this might be a better fit for how I've implemented this because I was overriding `zoekt.name` since I only knew how to filter by `r:` in my query strings.

When I'm trying to understand the searching code I'm only seeing [query string and some other unrelated options](https://github.com/sourcegraph/zoekt/blob/c301e5c82b6e1632dce5c39902691c359559852e/api.go#LL773) as parameters. Are the filters by id encoded in the query string somehow? For some time I was thinking that RPC code path was using the GPRC protocol and I've mostly been using `/search?format=json` but now I see I can use `/api/search` if I run with `-rpc`. Is this the API sourcegraph is using? If so is there some example request you could show for filtering by `repoid` in a performant way? It also makes me curious about whether I can just add more `zoekt.` metadata to the repos and search by that? For GitLab we have a hierachy of groups/subgroups/projects and it may be optimal for us to store the whole hierachy of ids in a field so that I can do a prefix search for efficiently searching whole groups at a time. I think this can be done using a regex as well but prefixes of ids seems more efficient for our usecase than regexes on the whole project name.

keegancsmith commented:
@DylanGriffith sorry my day has filled up. I'll get back to this before EOW (likely tomorrow).

DylanGriffith commented:
@keegancsmith I realise you're probably busy and there is no need to review this ASAP as the code is not yet complete but I did have 1 question you may be able to help with before I make more progress.

> If so for Sourcegraph we have added RPC interfaces, extended the index to take a integer ID as well as extending the query package to efficiently query for sets of IDs. We do a lot of the glue here in zoekt-sourcegraph-indexserver for further pointers. Hopefully that is enough to unblock ya.

I'm struggling to find the "query package" part of this statement. Where/how in the "query package" do you allow filtering by `zoekt.repoid`? From what I can see the only query syntax support is `repo:` which seems to expect a name. If you could point me to an MR or docs or just an example query that works I may be able to piece together the rest.

DylanGriffith committed:
Add /truncate endpoint to indexserver

stefanhengl commented:
> I'm struggling to find the "query package" part of this statement. Where/how in the "query package" do you allow filtering by zoekt.repoid? From what I can see the only query syntax support is repo: which seems to expect a name. If you could point me to an MR or docs or just an example query that works I may be able to piece together the rest.

Check out [BranchesRepos](https://sourcegraph.com/github.com/sourcegraph/zoekt/-/blob/query/query.go?L202:4&subtree=true) in query.go. [In Sourcegraph we construct the list of ids before sending it to Zoekt](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@adba3605ea75b655a14c5a1f6afc3b55716bb693/-/blob/internal/search/zoekt/indexed_search.go?L293)

DylanGriffith committed:
Switch to RepositoryId in zoekt-indexserver

DylanGriffith committed:
Make listen port dynamic for indexserver

keegancsmith commented on cmd/zoekt-git-clone/main.go:
```diff
@@ -45,16 +47,22 @@ func main() {
 	config := map[string]string{
-		"zoekt.name": name,
+		"zoekt.name":   name,
+		"zoekt.repoid": repoId,
```
I would only set this field if you have a non-empty ID. Additionally in other parts of Zoekt we assume the ID is a uint32. This is very sourcegraph specific and we can make it more generic if we want to make repoid a more first class citizen (which we do want in general)

keegancsmith commented on cmd/zoekt-indexserver/main.go:
```diff
@@ -164,6 +167,100 @@ func indexPendingRepos(indexDir, repoDir string, opts *Options, repos <-chan str
 }
 
+type indexRequest struct {
+	RepositoryUrl string // TODO: Decide if tokens can be in the URL or if we should pass separately
```
nit: I would call this CloneURL.

keegancsmith commented on cmd/zoekt-git-clone/main.go:
```diff
@@ -32,6 +32,8 @@ import (
 func main() {
 	dest := flag.String("dest", "", "destination directory")
+	nameFlag := flag.String("name", "", "name of repository")
+	repoIdFlag := flag.String("repoid", "", "id of repository")
```
nit: `repoIDFlag`

DylanGriffith replied:
Done

keegancsmith reviewed (approved):
Apologies again for the very slow turnaround time on review. Do not expect that in future, it is unusual. I have one or two minor comments inline, but this is pretty much ready to go.

DylanGriffith commented:
Thanks for the feedback @keegancsmith . We'll make some adjustments here and add specs before we send it back for another round of review. All your feedback makes a lot of sense.

DylanGriffith commented on cmd/zoekt-indexserver/main.go:
```diff
@@ -262,6 +359,8 @@ func main() {
 	dataDir := flag.String("data_dir",
 		filepath.Join(os.Getenv("HOME"), "zoekt-serving"), "directory holding all data.")
 	indexDir := flag.String("index_dir", "", "directory holding index shards. Defaults to $data_dir/index/")
+	dynamicIndexServer := flag.Bool("dynamic_server", false, "whether or not to start dynamic web server to allow client to specify repos to index dynamically")
```
Not necessary now that it's a separate binary.

DylanGriffith commented:
@keegancsmith this should be ready for another review now. We've made changes based on all of your feedback and there are also tests now.

keegancsmith commented on cmd/zoekt-dynamic-indexserver/main_test.go:
```diff
@@ -0,0 +1,96 @@
+
+var (
+	CmdTimeout = 100 * time.Millisecond
+)
```
minor: does not need to be exported.

dgruzd replied:
This is fixed 👍

keegancsmith commented on cmd/zoekt-git-clone/main.go:
```diff
@@ -32,6 +32,8 @@ import (
 func main() {
 	dest := flag.String("dest", "", "destination directory")
+	nameFlag := flag.String("name", "", "name of repository")
+	repoIDFlag := flag.String("repoid", "", "id of repository")
```
alternatively you can dissallow negative values / the zero value. Then use that as a signal it wasn't set.

dgruzd replied:
Thank you for the suggestion. We've changed it to [`flag.Uint`](https://github.com/sourcegraph/zoekt/pull/496/files#diff-f0a863bc5bc69c6afcbbb4d4d8e9c82f23d0b935116c062d9c5929932a13fc79R37) with 0 as the default value.

keegancsmith commented on cmd/zoekt-dynamic-indexserver/main_test.go:
```diff
@@ -0,0 +1,96 @@
+	expectedHistory := make([][]string, 3)
+	expectedHistory[0] = []string{"zoekt-git-clone", "-dest", "/repo_dir", "-name", "100", "-repoid", "100", "https://example.com/repository.git"}
+	expectedHistory[1] = []string{"git", "-C", "/repo_dir/100.git", "fetch"}
+	expectedHistory[2] = []string{"zoekt-git-index", "-index", "/index_dir", "/repo_dir/100.git"}
```
```suggestion
	expectedHistory := [][]string{
		{"zoekt-git-clone", "-dest", "/repo_dir", "-name", "100", "-repoid", "100", "https://example.com/repository.git"},
		{"git", "-C", "/repo_dir/100.git", "fetch"},
		{"zoekt-git-index", "-index", "/index_dir", "/repo_dir/100.git"},
}
```

keegancsmith commented on cmd/zoekt-dynamic-indexserver/main_test.go:
```diff
@@ -0,0 +1,96 @@
+	var buf bytes.Buffer
+	log.SetOutput(&buf)
+	f()
+	log.SetOutput(os.Stderr)
```
for mutating global state use defer to ensure you undo it in case of panic (eg log.Fatal uses a panic)
```suggestion
	log.SetOutput(&buf)
	defer func() { log.SetOutput(os.Stderr) }
	f()
```

keegancsmith commented on cmd/zoekt-dynamic-indexserver/main_test.go:
```diff
@@ -0,0 +1,96 @@
+	cmd := exec.CommandContext(ctx, "echo", "-n", "1")
+
+	stdout := captureOutput(func() {
+		out, err := loggedRun(cmd)
```
err is the idiomatic name for something with the type `error`. Rather call this var stderr.

keegancsmith commented on cmd/zoekt-dynamic-indexserver/main.go:
```diff
@@ -0,0 +1,212 @@
+		gitRepoPath,
+		"fetch",
+	}
+	executeCmd(ctx, "git", args...)
```
if this fails you still try to run zoekt-git-index

dgruzd replied:
We've added additional logic & tests to stop execution when one of the commands fails

keegancsmith commented on cmd/zoekt-dynamic-indexserver/main.go:
```diff
@@ -0,0 +1,212 @@
+	gitRepoPath, err := filepath.Abs(filepath.Join(opts.repoDir, fmt.Sprintf("%d.git", req.RepoID)))
+	if err != nil {
+		log.Printf("error loading git repo path: %v", err)
+		http.Error(w, "JSON parser error", http.StatusBadRequest)
```
this is not a parse error.

dgruzd replied:
This is now fixed

keegancsmith commented on cmd/zoekt-dynamic-indexserver/main.go:
```diff
@@ -0,0 +1,212 @@
+		gitRepoPath,
+	}
+	executeCmd(ctx, "zoekt-git-index", args...)
+}
```
you are not writing to w on success

dgruzd replied:
`serveIndex` now renders json on success https://github.com/sourcegraph/zoekt/pull/496/files#diff-d23d6eeb3bb1fdaf40622f2869e605ad254e69bac940b1e4a5b84fc590b3ada7R154

keegancsmith commented on cmd/zoekt-dynamic-indexserver/main.go:
```diff
@@ -0,0 +1,212 @@
+}
+
+func indexRepository(opts Options, req indexRequest, w http.ResponseWriter) {
+	ctx, cancel := context.WithTimeout(context.Background(), opts.indexTimeout)
```
I'm not sure why you are passing w in here. You don't write anything to it unless filepath.Abs fails (unlikely), and you don't respect the requests timeout (via r.GetContext()).

dgruzd replied:
I've removed `w` from `indexRepository`, thanks!

> and you don't respect the requests timeout (via r.GetContext()).

Do you mean that we should take `r.Context().Deadline()`? Right now `(deadline time.Time, ok bool)` returns `ok == false`, which means that we don't have one and will use `opts.indexTimeout`

dgruzd replied:
> However, you can use CommandContext to respect the ctx when running a command. Then if the user cancels the request it will cancel the index. But yeah, it likely makes sense to just ignore canceled requests to ensure you still index the repo.

I think ignoring cancelled requests makes more sense, thanks!

keegancsmith commented on cmd/zoekt-dynamic-indexserver/main.go:
```diff
@@ -0,0 +1,212 @@
+			cmd.Args, err, outBuf.String(), errBuf.String())
+	}
+
+	return outBuf.Bytes(), errBuf.Bytes()
```
why do you return these? you don't actually read the return values anywhere outside of tests.

dgruzd replied:
We were only using it for tests. I've removed them from `loggerRun`  [signature](https://github.com/sourcegraph/zoekt/pull/496/files#diff-d23d6eeb3bb1fdaf40622f2869e605ad254e69bac940b1e4a5b84fc590b3ada7R36), thank you!

dgruzd commented:
Thank you for great suggestions, @keegancsmith. I believe this PR should be ready for another look 🤝

keegancsmith commented on cmd/zoekt-dynamic-indexserver/main.go:
```diff
@@ -0,0 +1,256 @@
+		if _, err := os.Stat(s); err == nil {
+			continue
+		}
+
```
nit: if s is not a dir then this will work. FYI MkdirAll is a noop if s is already a dir. Just rely on that.
```suggestion
```

keegancsmith reviewed (approved):
I have a few inline comments, but they are quite minor so I am approving. With your go ahead I'll land, otherwise happy to land as is, review updates or review another PR.

dgruzd committed:
Add new cmd zoekt-dynamic-indexserver, add tests

dgruzd commented:
@keegancsmith thank you for the feedback. I've applied your suggestions, squashed commits, and fixed naked returns. I believe this should be ready for merge 🤝

DylanGriffith commented:
Thanks @keegancsmith for helping us get this through! Your feedback was great and collaborating on Zoekt is a lot of fun.
